### ISSUE TITLE
internal compiler error when calling paren-less `param` function with parentheses

### ISSUE THREAD

**Issue by lucaferranti** [chapel-lang/chapel#19945]
### Summary of Problem

I accidentally called a paren-less function (Math.INFINITY) with parentheses and that lead the following internal compiler error.

```
foo.chpl:1: internal error: RES-ALI-0241 chpl version 1.26.0
Note: This source location is a guess.

Internal errors indicate a bug in the Chapel compiler ("It's us, not you"),
and we're sorry for the hassle.  We would appreciate your reporting this bug --
please see https://chapel-lang.org/bugs.html for instructions.  In the meantime,
the filename + line number above may be useful in working around the issue.
```

### Steps to Reproduce

**Source Code:**
```chapel
writeln(INFINITY());
```

**Compile command:**

`chpl foo.chpl`

**Execution command:**

N/A because compile error

**Associated Future Test(s):**

### Configuration Information

- Output of `chpl --version`:
```
chpl version 1.26.0
  built with LLVM version 12.0.0
Copyright 2020-2022 Hewlett Packard Enterprise Development LP
Copyright 2004-2019 Cray Inc.
(See LICENSE file for more details)
```

- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
lferrant@LAPTOP-HGM2FD2S:~$ chapel-1.26.0/util/printchplenv --anonymize
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: llvm
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: bundled
CHPL_HWLOC: bundled
CHPL_RE2: bundled
CHPL_LLVM: system
CHPL_AUX_FILESYS: none
```

- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
```
gcc (Ubuntu 10.3.0-1ubuntu1) 10.3.0
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

- (For Cray systems only) Output of `module list`:

**Comment by dlongnecke-cray** [comment]
This program also reproduces:

```chapel
proc foo param { return 0; }

proc test() {
  writeln(foo());
}
test();
```

Some things I've noted:

- The function being scope resolved has to be marked `param`, because that way it gets folded away
- If the function is not marked `param`, and it returns e.g., an `int`, we'd get a `error: unresolved access of 'int(64)' by '()'` error instead

I'm not sure if the non-param error message is entirely correct/sensible either, but this specific bug only seems to happen for `param` functions. Somehow the combination of replacing a CallExpr's base expression with another CallExpr (making it a partial call) and then trying to param fold that after normalize makes everything explode.

Maybe the right thing to do is just to add a small little patch here in `resolveUnresolvedSymExpr` (in `scopeResolve`)?

```chpl
 940   } else if (fn->hasFlag(FLAG_NO_PARENS) == true) {
 941     checkIdInsideWithClause(usymExpr, usymExpr);
 942     usymExpr->replace(new CallExpr(fn));
 943
```

Where we just check to see if the target is marked as returning `param` and then emit a suitable error. Another suitable place to handle this could be `preFold`.

---

I am not sure what the error message should be, though.

**Comment by bradcray** [comment]
Thanks for taking a look at this David!  (and for filing it, Luca)

> I'm not sure if the non-param error message is entirely correct/sensible either

I'd call it accurate, but not necessarily clear to a user who thinks they're doing something appropriate.

> I am not sure what the error message should be, though.

If that works, then for now, I'd probably use the same wording as the non-param case for consistency (and then both can be improved when we continue getting questions about what it means).

One thing to be careful of here is that some param values can be indexed—notably (and, currently, maybe only?) strings and bytes.  So, for example, this program needs to continue working:

```chapel
proc foo param {
  return b"hi david";
}

proc bar param {
  return "hi david";
}

writeln(foo(0));
writeln(bar(0));
```

**Comment by lucaferranti** [comment]
what about something like

```
error: try to call parenthesis-less function $fun_name with parenthesis 
```

this would probably help the user (especially novice) debug what is causing it.

**Comment by dlongnecke-cray** [comment]
Actually, Brad's latest comment about how some of these calls should be possible at compile-time has made me think that maybe it's actually a bug with param folding. I think there's no particular reason why we shouldn't see that same error message at compile-time. Anyhow, before I go inserting error messages that potentially duplicate a lot of failure cases into scope resolve, I want to check param folding to see if there's something I can fix there.

**Comment by bradcray** [comment]
@lucaferranti:  I think that your suggestion:

```
error: trying to call parenthesis-less function $fun_name with parenthesis 
```

is an good proposal for us to try out as a different wording here.  Note that the reason for the current wording is that if the type returned by the paren-less function is itself indexable / callable, then it's not inherently problematic to apply parens to a paren-less function.  As an example, the following program is legal:

```chapel
proc foo {
  return [1, 2, 3];
}

writeln(foo(1)); 
```

and (I presume) at the time that we're realizing we can or can't apply the next set of arguments to the expression returned by the outer call, we've probably lost track of the context—whether it was a paren-less vs. paren-ful function, or some other expression.  That said, it probably wouldn't be too much work to look a bit further up the expression tree before emitting this error to specialize it for the case that it was a paren-less function call.

**Comment by dlongnecke-cray** [comment]
I have an initial fix which will change this from a compiler error to:

```
error: unresolved access of 'real(64)' by '()'
```

Which fits the interpretation of "you just tried to call a into real(64)". I think ideally we would mention the original callsite (the parenless method INFINITY in this example), but that might take a bit more work. I think dyno might be a bit less painful of a place to try making improvements, since it's much easier to do pattern matching on an immutable tree.

**Comment by bradcray** [comment]
Thanks for filing this @lucaferranti!  We didn't quite get it fixed in time for 1.27.0, but users of Chapel 1.28.0 and beyond will get the improved behavior.  (And thanks David for taking care of this!)
